tools/twix/src/repository_parameters.rs

```diff
@@ -32,46 +32,52 @@ impl RepositoryParameters {
 
     pub fn write(&self, address: &str, path: String, value: Value) {
         let repository = self.repository.clone();
-        let Ok(head_id) = self
-            .head_id_from_address(address)
+        let Ok(hardware_ids) = self
+            .hardware_ids_from_address(address)
         else {
             error!("failed to get head ID from address {address}");
             return
         };
         self.runtime.spawn(async move {
-            let mut stored_value = repository
-                .read_configuration(&head_id)
-                .await
-                .unwrap_or_default();
-
             let nested_value_to_be_added = nest_value_at_path(path.as_str(), &value);
 
-            merge_json(&mut stored_value, &nested_value_to_be_added);
+            let stored_value: Value = parameters::directory::deserialize(
+                repository.root_directory(),
+                &hardware_ids.body_id,
+                &hardware_ids.head_id,
+            )
+            .await
+            .unwrap_or_default();
+
+            let diff = get_diff_against_stored_value(&stored_value, &nested_value_to_be_added);
```

Comment:
What was the reason not to use `serialize()` from the configuration directory?

Reply:
Serialise directly writes to the location directory, whereas we generally want to write a level above.
I didn't want to modify the behaviour of serialise either as it is a critical component for deployment, etc parts.

Reply:
We need more functionality in the deserialization to select where to deserialize to (location, no location, head/body, ...). The communication interface would also benefit from this feature. @knoellle Do we want to make the serialization more intelligent in this PR?

Reply:
Yes this would remove some code duplication.

Reply:
I think that's a good idea, yes.
Parameter reading/writing should be implemented only once and then used both for tools running on dev machines as well as for code running on the nao.
I had already suggested unifying them in a previous PR during GORE but decided to postpone it.

Reply:
In that case, we can implement these functions into this PR and get it done.

My proposal for this:
- Decouple serialize function and write to the disk.
- Or indicate where to write by passing an argument to serialize

If this sounds reasonable, I can implement one of the solutions.

Reply:
I'm not sure what you mean by decoupling serialization from writing.
The actual serialization part is trivial and does not need to be extracted. On the other hand, merging of an updated branch or leaf with the current configuration and generating a diff already needs information about where the configuration will be saved to.

Thus, I argue for the second option.

Reply:
The second option also allows easy integration in the communication protocol.

Reply:
My current idea is to change the parameter directory API and make the API smarter by hiding complexity. Currently the serialization/deserialization are free functions and the idea is to store the parsed parameters with location information as state and convert the free functions into methods on this state. The goal of this design is to hide the location complexity inside of this new type.

The deserialization will be converted into a method that returns the flattened parameters (e.g. for usage in Communication).

For the serialization it will be possible to decide whether to add or remove a value at a path. Additions must contain the location (all locations, all locations with head ID, all locations with body ID, current location, current location with head ID, current location with body ID). Removals also must contain the location (it should also be possible to pass "don't care" which removes any value in any parameter location). With this design, we can more easily update parameter files and pass some hard decisions to the user (e.g. where to store the parameters).

This adjusted parameter directory API will be made available via Communication API und used in this PR for Twix parameter saving.

One drawback of this design is that the user can only get a flattened snapshot of the parameters but needs to provide location information while updating values. This may lead to situations where the user stores a value but a higher priority value at the same path still takes precedence resulting in no change in the flattened parameters. One mitigation for the user in this scenario is to first remove the value in all locations and then add it.

Open questions:

1. Only allow to store to the current location or to arbitrary locations?
2. Have a flat enum for the location decision or divide into multiple orthogonal enums e.g. whether it should be in a location and whether it should be default/head/body?

Reply:
I generally like the approach.

Regarding your questions:
1. Which "location" do you mean here? The term is overloaded by the physical location as well as the combination of physical location + head/body ID for the filed location.
2. I'd argue for having separate enums. Since they are - as you say - orthogonal.

I also have some questions of my own. I will continue your numbering to avoid confusion.

3. On the communication server side, would you keep this state in addition to what the server already stores or did you intend to replace it? In the former case the new state object would be in sync with the state on disk. That would have to advantage of being able to save parameters regardless of what has been temporarily set at runtime by a communication client.
4. In the "remove: don't care" case mentioned above, what do you mean by "any value in any location"? 
   1. Why "any value"? Would that delete the whole file at the specified location/all locations?
   2. "any location" meaning the `Location` enum above or across all physical locations?

Reply:
1. I meant the combination of location and IDs. The question is if we want to limit the user that it will be all or the current location (regardless of IDs) or we allow that it can be all or a given location (e.g. by string, regardless of IDs).
2. ACK
3. We currently have this buffering semantic and we can keep it. The buffering would be in the form of the state within the new type.
4. The idea with "any" is that we delete all values (and children) at a given path in all combinations of locations and IDs. This would be for user convenience that the user does not need to trigger removal calls for each combination but rather the server iterates and deletes everything under a given path for all combinations.
5. I just noticed in 4. that with the removal feature, we could produce invalid configurations because we don't "default" the value at the path but completely nuke it. If there is a bug and we are not adding something back, this will be a problem. I'm not sure how to mitigate this. 🤔

Reply:
The whole idea of an incremental update API will only work for the Twix case because there we are directly writing JSON values into the JSON object. But for communication, this will not work because there we are using `SerializeHierarchy`. `SerializeHierarchy` works on the flattened JSON object and has no knowledge about locations and IDs. The proposed parameter directory API does not help with that because we need to somehow recover location information.

New idea similar what Darshana already started: We could diff the stored parameters on disk with the flattened parameters in RAM. This diff is then stored to the combination of location and ID that the user requests.

One scenario could be that we manually modify the calibration parameters via Twix in the Communication server. Afterwards we are storing the parameters to disk, i.e., the diff results in the calibration parameters and the user specified to store this diff in the root head parameters. Should work.

Another scenario could be that we tweak some vision parameters and also the calibration parameters. We would like to store the vision parameters in the root parameters for all robots but the calibration parameters only for the particular robot. A solution would be to let the user specify the path to store to disk in addition to the combination of location and IDs.

Btw, let's call the combination of location and IDs from now on *scope*.

This would result in the following API:

```rust
pub async fn deserialize<Parameters>(
    parameters_root_path: impl AsRef<Path>,
    body_id: &str,
    head_id: &str,
) -> Result<Parameters, DirectoryError>
where
    Parameters: DeserializeOwned,
{}

pub async fn serialize<Parameters>(
    parameters: &Parameters,
    scope: Scope,  // <-------------------------- new
    parameters_root_path: impl AsRef<Path>,
    body_id: &str,
    head_id: &str,
) -> Result<(), DirectoryError>
where
    Parameters: Serialize,
{}

struct Scope {
    location: Location,
    id: Id,
    path: String, // <----- path in the JSON diff object to store in this serialization
}                 //        will be similar to the path used in SerializeHierarchy
enum Location { All, Current }
enum Id { All, Body, Head }
```

Reply:
Works as planned: https://github.com/HULKs/hulk/pull/267/commits/9a851f668f3d9e7efa1c5f5cb74547177591a1cd

Reply:
You are correct. Not done yet.

Reply:
Now integrated into Twix and repository crate.